File: testdata/ts/simple.ts
import { EventEmitter } from 'events';

interface Greeter {
    greet(name: string): string;
    farewell(name: string): string;
}

interface Logger {
    log(message: string): void;
}

class ConsoleGreeter implements Greeter, Logger {
    greet(name: string): string {
        return `Hello, ${name}!`;
    }

    farewell(name: string): string {
        return `Goodbye, ${name}!`;
    }

    log(message: string): void {
        console.log(message);
    }
}

class VerboseGreeter extends ConsoleGreeter {
    greet(name: string): string {
        this.log(`Greeting ${name}`);
        return super.greet(name);
    }
}

function createGreeter(): Greeter {
    return new ConsoleGreeter();
}

const greeter = createGreeter();
greeter.greet('world');

File: testdata/objc/simple.m
#import <Foundation/Foundation.h>
#import "Greeter.h"
#include "utils.h"

@protocol GreeterProtocol
- (NSString *)greet:(NSString *)name;
@end

@interface SimpleGreeter : NSObject <GreeterProtocol>
@property (nonatomic, strong) NSString *prefix;
@end

@implementation SimpleGreeter

- (instancetype)initWithPrefix:(NSString *)prefix {
    self = [super init];
    if (self) {
        _prefix = prefix;
    }
    return self;
}

- (NSString *)greet:(NSString *)name {
    NSString *formatted = [self formatName:name];
    return [NSString stringWithFormat:@"%@ %@!", self.prefix, formatted];
}

- (NSString *)formatName:(NSString *)name {
    return [name capitalizedString];
}

@end

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        SimpleGreeter *greeter = [[SimpleGreeter alloc] initWithPrefix:@"Hello"];
        NSString *result = [greeter greet:@"world"];
        NSLog(@"%@", result);
    }
    return 0;
}
